proposal/js-deprecations.d.ts

```diff
@@ -0,0 +1,302 @@
+        span?: SourceSpan;
+        stack?: string;
+      }
+    ): void;
+  }
+
+  /**
+   * An object containing all of the deprecations recognized by this
+   * implementation as properties, as well as a `find` function to get the
+   * deprecation for a given ID.
+   */
+  export const deprecations: Readonly<
+    _Deprecations & {[key: string]: Deprecation}
+  >;
+}
+
+interface _Deprecations {
+  /** Deprecation for passing a string to `call` instead of `get-function`. */
+  callString: Deprecation<'call-string', 'active'>;
+
+  /** Deprecation for `@elseif`. */
+  elseif: Deprecation<'elseif', 'active'>;
+
+  /** Deprecation for parsing `@-moz-document`. */
+  mozDocument: Deprecation<'moz-document', 'active'>;
+
+  /** Deprecation for importers using relative canonical URLs. */
+  relativeCanonical: Deprecation<'relative-canonical', 'active'>;
+
+  /** Deprecation for declaring new variables with `!global`. */
+  newGlobal: Deprecation<'new-global', 'active'>;
+
+  /**
+   * Deprecation for certain functions in the color module matching the
+   * behavior of their global counterparts for compatibility reasons.
+   */
+  colorModuleCompat: Deprecation<'color-module-compat', 'active'>;
+
+  /** Deprecation for treaing `/` as division. */
+  slashDiv: Deprecation<'slash-div', 'active'>;
+
+  /** Deprecation for leading, trailing, and repeated combinators. */
+  bogusCombinators: Deprecation<'bogus-combinators', 'active'>;
+
+  /** Deprecation for ambiguous `+` and `-` operators. */
+  strictUnary: Deprecation<'strict-unary', 'active'>;
+
+  /** Deprecation for passing invalid units to certain built-in functions. */
+  functionUnits: Deprecation<'function-units', 'active'>;
+
+  /** Deprecation for `@import` rules. */
+  importRules: Deprecation<'import', 'future'>;
+
+  /** Used for deprecations coming from user-authored code. */
+  userAuthored: Deprecation<'user-authored', 'user'>;
+
+  /**
+   * Returns the deprecation with the given ID, or null if none exists.
+   *
+   * This must support all specified deprecations, but implementations may also
+   * have their own deprecations beyond those specified by the language.
+   */
+  find(id: string): Deprecation<typeof id, DeprecationStatus> | null;
+}
+
+/** A deprecation, or the ID of one. */
+export type DeprecationOrId = Deprecation | string;
+
+/** A deprecation's status. */
+export type DeprecationStatus = 'active' | 'user' | 'future' | 'obsolete';
+
+/** A deprecated feature in the language. */
+export interface Deprecation<
+  id extends string = string,
+  status extends DeprecationStatus = DeprecationStatus
+> {
+  /**
+   * An ID for this deprecation.
+   *
+   * For language-wide deprecations specified in the API, this will be a short,
+   * alphanumeric, kebab-case identifier. Implementation-specific deprecations
+   * should be prefixed with an implementation identifier followed by a slash to
+   * avoid collision with future language-wide deprecations.
```

Comment:
I'm not sure how feasible it is to distinguish impl-specific deprecations by prefix. Currently _all_ deprecations are essentially Dart Sass only, since there are no other substantially-compatible impls. Even if there were, which impls actually need to deprecate behavior can be contingent on whether/when they implement the original behavior—when we were first rolling out Dart Sass, we just implemented the new non-deprecated behaviors for a number of "standard" Ruby Sass deprecations.

I think generally it's probably a good idea to just have all deprecations share a namespace, and ask implementation authors to collaborate on a central set of IDs. I can't imagine there will ever be so many implementations of such independence that that becomes onerous.